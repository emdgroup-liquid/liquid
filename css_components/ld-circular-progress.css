.ld-circular-progress,:host{--ld-circular-progress-valuemin:0;--ld-circular-progress-valuemax:100;--ld-circular-progress-valuenow:0;--ld-circular-progress-size:7.5rem;--ld-circular-progress-transition-duration:0.2s;--ld-circular-progress-stroke-mask:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 200 200"><circle cx="100" cy="100" r="89" stroke="red" stroke-width="20"/></svg>');--ld-circular-progress-bar-col:var(--ld-thm-primary);--ld-circular-progress-bar-col-overflow:var(--ld-thm-error);--ld-circular-progress-bg-col:var(--ld-col-neutral-100);--ld-circular-progress-bg-col-overflow:var(--ld-thm-error);--ld-circular-progress-circle-bg-col:var(--ld-circular-progress-bg-col);--ld-circular-progress-circle-bg-col-overflow:var(
    --ld-circular-progress-bg-col-overflow
  );--ld-circular-progress-pi:3.141592653589793;--ld-circular-progress-wht-overlay-col:var(--ld-col-wht-alpha-high);--ld-circular-progress-calc-valuenow:max(0,calc(var(--ld-circular-progress-valuenow) - var(--ld-circular-progress-valuemin)));--ld-circular-progress-calc-valuemax:calc(var(--ld-circular-progress-valuemax) - var(--ld-circular-progress-valuemin));--ld-circular-progress-calc-relative-progress:calc(var(--ld-circular-progress-calc-valuenow)/var(--ld-circular-progress-calc-valuemax));--ld-circular-progress-has-overflow:clamp(0,calc((var(--ld-circular-progress-calc-relative-progress) - 1)*99999999999),1);align-items:center;display:inline-flex;flex-direction:column;gap:var(--ld-sp-4);height:calc(var(--ld-circular-progress-size) - var(--ld-circular-progress-size)*.4);justify-content:center;max-width:100%;padding:calc(var(--ld-circular-progress-size)*.165) calc(var(--ld-circular-progress-size)*.2) calc(var(--ld-circular-progress-size)*.235);position:relative;width:calc(var(--ld-circular-progress-size) - var(--ld-circular-progress-size)*.4)}.ld-circular-progress:after,.ld-circular-progress:before,:host:after,:host:before{content:"";inset:0;-webkit-mask-image:var(--ld-circular-progress-stroke-mask);mask-image:var(--ld-circular-progress-stroke-mask);position:absolute;transition:opacity var(--ld-circular-progress-transition-duration) linear}.ld-circular-progress:before,:host:before{background-color:var(--ld-circular-progress-bg-col);opacity:calc(1 - var(--ld-circular-progress-has-overflow))}.ld-circular-progress:after,:host:after{box-shadow:inset 0 0 0 calc(var(--ld-circular-progress-has-overflow)*99rem) var(--ld-circular-progress-wht-overlay-col),inset 0 0 0 calc(var(--ld-circular-progress-has-overflow)*99rem) var(--ld-circular-progress-bg-col-overflow);opacity:var(--ld-circular-progress-has-overflow);position:absolute}.ld-circular-progress--brand-color,:host(.ld-circular-progress--brand-color){--ld-circular-progress-bar-col:var(--ld-col-wht);--ld-circular-progress-bg-col:var(--ld-thm-primary-focus);--ld-circular-progress-bg-col-overflow:var(--ld-thm-error-active);--ld-circular-progress-wht-overlay-col:var(--ld-col-wht-alpha-low)}.ld-circular-progress__stroke{fill:none;inset:0;-webkit-mask-image:var(--ld-circular-progress-stroke-mask);mask-image:var(--ld-circular-progress-stroke-mask);position:absolute;z-index:1}.ld-circular-progress__stroke circle{stroke-width:2.2;stroke-dasharray:calc(100/var(--ld-circular-progress-pi));transform:scaleX(-1) rotate(-90deg);transform-origin:center;transition:opacity var(--ld-circular-progress-transition-duration) linear,stroke-dashoffset var(--ld-circular-progress-transition-duration) ease}.ld-circular-progress__stroke circle:first-of-type{stroke:var(--ld-circular-progress-bar-col);stroke-dashoffset:min(0,calc((-99 + var(--ld-circular-progress-calc-relative-progress)*100)/var(--ld-circular-progress-pi)));opacity:calc(1 - var(--ld-circular-progress-has-overflow))}.ld-circular-progress__stroke circle:last-of-type{stroke:var(--ld-circular-progress-bar-col-overflow);stroke-dashoffset:max(calc(-100/var(--ld-circular-progress-pi)),min(0,calc((-99 + (var(--ld-circular-progress-calc-relative-progress) - 1)*100)/var(--ld-circular-progress-pi))));opacity:var(--ld-circular-progress-has-overflow)}