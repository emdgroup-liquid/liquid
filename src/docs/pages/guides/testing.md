---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

