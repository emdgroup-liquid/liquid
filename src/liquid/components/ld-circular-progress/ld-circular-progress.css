:host,
.ld-circular-progress {
  --ld-circular-progress-valuemin: 0;
  --ld-circular-progress-valuemax: 100;
  --ld-circular-progress-valuenow: 0;
  --ld-circular-progress-size: 7.5rem;
  --ld-circular-progress-transition-duration: 0.2s;
  --ld-circular-progress-stroke-mask: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 200 200"><circle cx="100" cy="100" r="89" stroke="red" stroke-width="20"/></svg>');

  /* colors */
  --ld-circular-progress-bar-col: var(--ld-thm-primary);
  --ld-circular-progress-bar-col-overflow: var(--ld-thm-error);
  --ld-circular-progress-bg-col: var(--ld-col-neutral-100);
  --ld-circular-progress-bg-col-overflow: var(--ld-thm-error);
  --ld-circular-progress-circle-bg-col: var(--ld-circular-progress-bg-col);
  --ld-circular-progress-circle-bg-col-overflow: var(
    --ld-circular-progress-bg-col-overflow
  );
  --ld-circular-progress-pi: 3.141592653589793;
  --ld-circular-progress-wht-overlay-col: var(--ld-col-wht-alpha-high);

  /* computed */
  --ld-circular-progress-calc-valuenow: max(
    0,
    calc(
      var(--ld-circular-progress-valuenow) -
        var(--ld-circular-progress-valuemin)
    )
  );
  --ld-circular-progress-calc-valuemax: calc(
    var(--ld-circular-progress-valuemax) - var(--ld-circular-progress-valuemin)
  );
  --ld-circular-progress-calc-relative-progress: calc(
    var(--ld-circular-progress-calc-valuenow) /
      var(--ld-circular-progress-calc-valuemax)
  );
  --ld-circular-progress-has-overflow: clamp(
    0,
    calc(
      (var(--ld-circular-progress-calc-relative-progress) - 1) * 99999999999
    ),
    1
  );

  align-items: center;
  display: inline-flex;
  flex-direction: column;
  gap: var(--ld-sp-4);
  height: calc(
    var(--ld-circular-progress-size) - var(--ld-circular-progress-size) * 0.4
  );
  max-width: 100%;
  justify-content: center;
  padding: calc(var(--ld-circular-progress-size) * 0.165)
    calc(var(--ld-circular-progress-size) * 0.2)
    calc(var(--ld-circular-progress-size) * 0.235);
  position: relative;
  width: calc(
    var(--ld-circular-progress-size) - var(--ld-circular-progress-size) * 0.4
  );

  &::before,
  &::after {
    content: '';
    inset: 0;
    mask-image: var(--ld-circular-progress-stroke-mask);
    position: absolute;
    transition: opacity var(--ld-circular-progress-transition-duration) linear;
  }

  &::before {
    background-color: var(--ld-circular-progress-bg-col);
    opacity: calc(1 - var(--ld-circular-progress-has-overflow));
  }

  &::after {
    box-shadow: inset 0px 0px 0px
        calc(var(--ld-circular-progress-has-overflow) * 99rem)
        var(--ld-circular-progress-wht-overlay-col),
      inset 0px 0px 0px calc(var(--ld-circular-progress-has-overflow) * 99rem)
        var(--ld-circular-progress-bg-col-overflow);
    opacity: var(--ld-circular-progress-has-overflow);
    position: absolute;
  }
}

:host(.ld-circular-progress--brand-color),
.ld-circular-progress--brand-color {
  --ld-circular-progress-bar-col: var(--ld-col-wht);
  --ld-circular-progress-bg-col: var(--ld-thm-primary-focus);
  --ld-circular-progress-bg-col-overflow: var(--ld-thm-error-active);
  --ld-circular-progress-wht-overlay-col: var(--ld-col-wht-alpha-low);
}

.ld-circular-progress__stroke {
  position: absolute;
  inset: 0;
  fill: none;
  mask-image: var(--ld-circular-progress-stroke-mask);
  z-index: 1;

  circle {
    transform-origin: center;
    stroke-width: 2.2;
    stroke-dasharray: calc(100 / var(--ld-circular-progress-pi));
    transition: opacity var(--ld-circular-progress-transition-duration) linear,
      stroke-dashoffset var(--ld-circular-progress-transition-duration) ease;
    transform: scaleX(-1) rotate(-90deg);

    &:first-of-type {
      stroke: var(--ld-circular-progress-bar-col);
      stroke-dashoffset: min(
        0,
        calc(
          (-100 + var(--ld-circular-progress-calc-relative-progress) * 100 + 1) /
            var(--ld-circular-progress-pi)
        )
      );
      opacity: calc(1 - var(--ld-circular-progress-has-overflow));
    }
    &:last-of-type {
      stroke: var(--ld-circular-progress-bar-col-overflow);
      stroke-dashoffset: max(
        calc(-100 / var(--ld-circular-progress-pi)),
        min(
          0,
          calc(
            (
                -100 + (var(--ld-circular-progress-calc-relative-progress) - 1) *
                  100 + 1
              ) / var(--ld-circular-progress-pi)
          )
        )
      );
      opacity: var(--ld-circular-progress-has-overflow);
    }
  }
}
