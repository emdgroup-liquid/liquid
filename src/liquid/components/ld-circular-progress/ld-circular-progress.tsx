import { Component, h, Host, Prop } from '@stencil/core'
import { getClassNames } from '../../utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-circular-progress',
  styleUrl: 'ld-circular-progress.css',
  shadow: true,
})
export class LdCircularProgress {
  /**
   * Defines the string value or identifies the element (or elements)
   * that label the progressbar element providing an accessible name.
   */
  @Prop({ reflect: true }) ariaLabeledby?: string

  /**
   * Set to a decimal value representing the maximum value, and greater
   * than aria-valuemin. If not present, the default value is 100.
   */
  @Prop({ reflect: true }) ariaValuemax? = 100

  /**
   * Set to a decimal value representing the minimum value, and less
   * than aria-valuemax. If not present, the default value is 0.
   */
  @Prop({ reflect: true }) ariaValuemin? = 0

  /**
   * Only present and required if the value is not indeterminate.
   * Set to a decimal value between 0, or valuemin if present,
   * and aria-valuemax indicating the current value of the progress bar.
   */
  @Prop({ reflect: true }) ariaValuenow?: number

  /**
   * Assistive technologies often present the value of aria-valuenow
   * as a percentage. If this would not be accurate use this property
   * to make the progress bar value understandable.
   */
  @Prop({ reflect: true }) ariaValuetext?: string

  // `onBrandColor` is not possible: Stencil expects `on*` props to be event handlers.
  /**
   * Styles the progress bar in a way that it looks good on the
   * primary color of the current theme.
   */
  @Prop() brandColor?: boolean

  render() {
    const cl = getClassNames([
      'ld-circular-progress',
      this.brandColor && 'ld-circular-progress--brand-color',
    ])

    return (
      <Host
        class={cl}
        role="progressbar"
        style={{
          ...(this.ariaValuemax !== undefined && {
            '--ld-circular-progress-valuemax': this.ariaValuemax + '',
          }),
          ...(this.ariaValuemin !== undefined && {
            '--ld-circular-progress-valuemin': this.ariaValuemin + '',
          }),
          ...(this.ariaValuenow !== undefined && {
            '--ld-circular-progress-valuenow': this.ariaValuenow + '',
          }),
        }}
      >
        <svg class="ld-circular-progress__stroke" viewBox="0 0 10 10">
          <circle cx="5" cy="5" r="5" />
          <circle cx="5" cy="5" r="5" />
        </svg>
        <slot></slot>
      </Host>
    )
  }
}
