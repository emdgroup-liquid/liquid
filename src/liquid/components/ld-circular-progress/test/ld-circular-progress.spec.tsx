import { newSpecPage } from '@stencil/core/testing'
import { LdCircularProgress } from '../ld-circular-progress'
import { h } from '@stencil/core'

describe('ld-circular-progress', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdCircularProgress],
      template: () => <ld-circular-progress aria-valuenow="25" />,
    })
    expect(page.root).toMatchSnapshot()
  })

  describe('adds classes according to props', () => {
    it('adds brand-color class', async () => {
      const page = await newSpecPage({
        components: [LdCircularProgress],
        template: () => <ld-circular-progress aria-valuenow="25" brand-color />,
      })
      expect(page.root).toHaveClass('ld-circular-progress--brand-color')
    })
  })

  describe('applies css custom props according to props', () => {
    it('applies valuemax prop', async () => {
      const page = await newSpecPage({
        components: [LdCircularProgress],
        template: () => <ld-circular-progress aria-valuemax="200" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('applies valuemin prop', async () => {
      const page = await newSpecPage({
        components: [LdCircularProgress],
        template: () => <ld-circular-progress aria-valuemin="10" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('applies valuenow prop', async () => {
      const page = await newSpecPage({
        components: [LdCircularProgress],
        template: () => <ld-circular-progress aria-valuenow="25" />,
      })
      expect(page.root).toMatchSnapshot()
    })
  })
})
