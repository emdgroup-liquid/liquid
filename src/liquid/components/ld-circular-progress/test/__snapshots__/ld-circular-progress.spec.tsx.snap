// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-circular-progress applies css custom props according to props applies valuemax prop 1`] = `
<ld-circular-progress aria-valuemax="200" aria-valuemin="0" class="ld-circular-progress" role="progressbar" style="--ld-circular-progress-valuemax: 200; --ld-circular-progress-valuemin: 0;">
  <mock:shadow-root>
    <svg class="ld-circular-progress__stroke" viewBox="0 0 10 10">
      <circle cx="5" cy="5" r="5"></circle>
      <circle cx="5" cy="5" r="5"></circle>
    </svg>
    <slot></slot>
  </mock:shadow-root>
</ld-circular-progress>
`;

exports[`ld-circular-progress applies css custom props according to props applies valuemin prop 1`] = `
<ld-circular-progress aria-valuemax="100" aria-valuemin="10" class="ld-circular-progress" role="progressbar" style="--ld-circular-progress-valuemax: 100; --ld-circular-progress-valuemin: 10;">
  <mock:shadow-root>
    <svg class="ld-circular-progress__stroke" viewBox="0 0 10 10">
      <circle cx="5" cy="5" r="5"></circle>
      <circle cx="5" cy="5" r="5"></circle>
    </svg>
    <slot></slot>
  </mock:shadow-root>
</ld-circular-progress>
`;

exports[`ld-circular-progress applies css custom props according to props applies valuenow prop 1`] = `
<ld-circular-progress aria-valuemax="100" aria-valuemin="0" aria-valuenow="25" class="ld-circular-progress" role="progressbar" style="--ld-circular-progress-valuemax: 100; --ld-circular-progress-valuemin: 0; --ld-circular-progress-valuenow: 25;">
  <mock:shadow-root>
    <svg class="ld-circular-progress__stroke" viewBox="0 0 10 10">
      <circle cx="5" cy="5" r="5"></circle>
      <circle cx="5" cy="5" r="5"></circle>
    </svg>
    <slot></slot>
  </mock:shadow-root>
</ld-circular-progress>
`;

exports[`ld-circular-progress renders 1`] = `
<ld-circular-progress aria-valuemax="100" aria-valuemin="0" aria-valuenow="25" class="ld-circular-progress" role="progressbar" style="--ld-circular-progress-valuemax: 100; --ld-circular-progress-valuemin: 0; --ld-circular-progress-valuenow: 25;">
  <mock:shadow-root>
    <svg class="ld-circular-progress__stroke" viewBox="0 0 10 10">
      <circle cx="5" cy="5" r="5"></circle>
      <circle cx="5" cy="5" r="5"></circle>
    </svg>
    <slot></slot>
  </mock:shadow-root>
</ld-circular-progress>
`;
